### PR TITLE
Remove allocation from ResourceLeakDetector

### PR DESCRIPTION
**Pull request by carl-mastrangelo**
Motivation:
RLD allocates an ArrayDeque in anticipation of recording access
points.  If the leak detection level is less than ADVANCED though,
the dequeue is never used.  Since SIMPLE is the default level,
there is a minor perf win to not preemptively allocate it.

This showed up in garbage profiling when creation a high number of
buffers.

Modifications:
Only allocate the dequeue if it will be used.

Result:
Less garbage created.

### CODE REVIEW

File: common/src/main/java/io/netty/util/ResourceLeakDetector.java [hunk 1]
```diff
@@ -407,9 +408,14 @@ public String toString() {
...
-                array = lastRecords.toArray();
-                removedRecords = this.removedRecords;
+            if (lastRecords != null) {
+                synchronized (lastRecords) {
+                    array = lastRecords.toArray();
+                    removedRecords = this.removedRecords;
+                }
+            } else {
+                removedRecords = 0;
+                array = new Object[0];
```
Comment (normanmaurer) [inline common/src/main/java/io/netty/util/ResourceLeakDetector.java]: @carl-mastrangelo use `EmptyArrays.EMPTY_OBJECTS` to reduce allocations ;)

Comment (carl-mastrangelo) [reply common/src/main/java/io/netty/util/ResourceLeakDetector.java]: Done.

**Comment by normanmaurer** [comment]
Cherry-picked into 4.1 (b32cd26a961c71cc5ad0e9f88fa8c15fbfd89930) and 4.0 (9df8082c39050f206780ab81e96624e3801f418f)

**Comment by normanmaurer** [comment]
@carl-mastrangelo thanks!

**Pull request closed without merging**